src/background/StreamSessionManager.ts

```diff
@@ -0,0 +1,373 @@
+    private peer = new Peer();
+
+    /** The container for all session state information */
+    private sessionState: ISessionState = {
+        leaderId: "",
+        sessionJoinOrder: [],
+        sessionUsers: new Map<string, IUserConnection>()
+    };
+
+    /** The current users connection state */
+    private connectionState: ConnectionState = ConnectionState.Disconnected;
+
+    public constructor(port: chrome.runtime.Port) {
+        this.port = port;
+
+        this.peer.on("open", (id: string) => {
+            this.isReady = true;
+            this.currentUserId = id;
+
+            this.sessionState.sessionUsers.set(id, {
+                id,
+                displayName: "",
+                permissions: {
+                    canControlVideo: true
+                }
+            });
+
+            console.warn("ctor - updated session state");
+            console.warn(this.sessionState);
+        });
+
+        port.onMessage.addListener(message => {
+            if (!this.currentUserId) {
+                return;
+            }
+
+            const currentUser = this.sessionState.sessionUsers.get(this.currentUserId);
+            if (currentUser?.permissions.canControlVideo) {
+                this.sessionState.sessionUsers.forEach(user => {
+                    if (user.id !== this.currentUserId) {
+                        user.connection?.send(message);
+                    }
+                });
+            }
+        });
+
+        this.peer.on("connection", (conn: Peer.DataConnection) => {
+            this.onNewConnection(conn);
+        });
+    }
+
+    /**
+     * Initializes the state if the current user is leader of the current session
+     */
+    public initializeStateAsSessionStarter(): void {
+        console.warn("Init as session starter");
+        this.connectionState = ConnectionState.Connected;
+        this.sessionState = {
+            leaderId: this.currentUserId!,
+            sessionJoinOrder: [this.currentUserId!],
+            sessionUsers: this.sessionState.sessionUsers
+        };
+        console.warn(this.sessionState);
+    }
+
+    /**
+     * Connects to a a peer. Can be used to start a session join if the current ConnectionState is Disconnected
+     * @param peerId The id of the peer to initially connect to
+     * @param [isReferral=false] Optional param that indicates that the peer we're connecting to is a referral
+     */
+    public connectToPeer(peerId: string, isReferral = false): void {
+        console.warn(`connect to peer: ${peerId} isReferral=${isReferral}`);
+        if (isReferral) {
+            this.connectionState = ConnectionState.ConnectingToPeer;
+            this.connectToExistingSessionReferralId = peerId;
+        }
+
+        const connectionMetadata: IConnectToPeerMetadata = {
+            // Todo - pass user display name (stored in localstorage?)
+            userDisplayName: "",
+            connectionState: this.connectionState
```

Comment:
Sure that works

Reply:
Actually I think I prefer connectionState because we might need it in the future if we're getting messages while we're adding the user to the topology (ConnectingToSession). Also think it's a good model of the current state of the session lifecycle for the current user